sbt-plugin/src/main/scala-sbt-1.0/org/scalajs/sbtplugin/SJSIRFileManager.scala

```diff
@@ -0,0 +1,100 @@
+/*
+ * Scala.js (https://www.scala-js.org/)
+ *
+ * Copyright EPFL.
+ *
+ * Licensed under Apache License 2.0
+ * (https://www.apache.org/licenses/LICENSE-2.0).
+ *
+ * See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.
+ */
+
+package org.scalajs.sbtplugin
+
+import java.io.File
+import java.nio.file.Files
+
+import scala.collection.mutable
+
+import sbt._
+
+import xsbti.compile.ClassFileManager
+
+/** A class file manager that prunes .sjsir files as needed.
+ *
+ *  This makes sure that, when a .class file must be deleted, the
+ *  corresponding .sjsir file is also deleted, as well as maintaining the
+ *  transactional aspect of incremental compilation in the presence of
+ *  compilation errors in an intermediate run (as product files need to be
+ *  backed up and later restored).
+ *
+ *  This code is adapted from Zinc `TransactionalClassFileManager`.
+ *  We need to duplicate the logic since forwarding to the default class
+ *  file manager doesn't work: we need to backup sjsir files in a different
+ *  temporary directory as class files.
+ *
+ *  We stole this from `TastyFileManager` in dotty, at
+ *  https://github.com/lampepfl/dotty/blob/0.17.0-RC1/sbt-dotty/src/dotty/tools/sbtplugin/TastyFileManager.scala
+ *
+ *  There are plans to upstream a generic version of this code in zinc in the
+ *  future: https://github.com/sbt/zinc/issues/579
+ */
+private[sbtplugin] final class SJSIRFileManager extends ClassFileManager {
+  private[this] var _tempDir: File = null
+
+  private[this] def tempDir: File = {
+    if (_tempDir == null) {
+      _tempDir = Files.createTempDirectory("backup").toFile
+    }
+    _tempDir
+  }
+
+  private[this] val generatedSJSIRFiles = new mutable.HashSet[File]
+  private[this] val movedSJSIRFiles = new mutable.HashMap[File, File]
+
+  override def delete(classes: Array[File]): Unit = {
+    val files = sjsirFiles(classes)
+    val toBeBackedUp = files
+      .filter(t => t.exists && !movedSJSIRFiles.contains(t) && !generatedSJSIRFiles(t))
```

Comment:
The existence test is redundant: `sjsirFiles` already only returns files that exist.

Reply:
When is the test for generatedSJSIRFiles relevant? From what I understand, if it ever triggers, this would be a bug no? Because a file deleted prior to one compilation step would be deleted prior to another compilation step of the same run...

Reply:
When is the test for `generatedSJSIRFiles` relevant? From what I understand, if it ever triggers, this would be a bug no? Because a file generated in one compilation step would be deleted in another compilation step of the same run...

Note: I'm following the doc here: https://github.com/sbt/zinc/blob/develop/internal/compiler-interface/src/main/java/xsbti/compile/ClassFileManager.java

Reply:
OK I figured it out. It is actually necessary. The thing is that incremental compilation works in a loop: each partial compilation can find out that more sources should be recompiled, and because of cycles, it happens that a same source file will be recompiled twice or more in the same cycle. If something has been generated by a first iteration, but deleted in the second iteration, we must not back it up. It must stay deleted at the end of the run (even when `!success`).

In zinc, we find

1. [`Incremental.compile`](https://github.com/sbt/zinc/blob/9200d1e7ec363fc069aa0fc903f548b786410706/internal/zinc-core/src/main/scala/sbt/internal/inc/Incremental.scala#L922) calls `cycle` inside `manageClassfiles`
2. [`manageClassfiles`](https://github.com/sbt/zinc/blob/9200d1e7ec363fc069aa0fc903f548b786410706/internal/zinc-core/src/main/scala/sbt/internal/inc/Incremental.scala#L149) performs a `try..catch` to always call `success` at the end (with `false` if it caught an exception)
3. [`cycle`](https://github.com/sbt/zinc/blob/9200d1e7ec363fc069aa0fc903f548b786410706/internal/zinc-core/src/main/scala/sbt/internal/inc/IncrementalCommon.scala#L762) is a loop that calls `recompileClasses` during one opération.
4. [`recompileClasses`](https://github.com/sbt/zinc/blob/9200d1e7ec363fc069aa0fc903f548b786410706/internal/zinc-core/src/main/scala/sbt/internal/inc/IncrementalCommon.scala#L170) calls `delete` once then `generated` once

So we can definitely receive a sequence of calls where `delete` then `generated` are called several times before a single `success` is called.